To perform limb detection, we will be using a modified MPII Human Pose dataset. First, we'll need to download the entire dataset. This file seeks to filter and shrink this dataset into a smaller subset for easier use. This does not update the dataset to have bounding boxes either.

In [ ]:
!wget https://datasets.d2.mpi-inf.mpg.de/andriluka14cvpr/mpii_human_pose_v1.tar.gz

--2021-03-27 19:24:32--  https://datasets.d2.mpi-inf.mpg.de/andriluka14cvpr/mpii_human_pose_v1.tar.gz
Resolving datasets.d2.mpi-inf.mpg.de (datasets.d2.mpi-inf.mpg.de)... 139.19.206.177
Connecting to datasets.d2.mpi-inf.mpg.de (datasets.d2.mpi-inf.mpg.de)|139.19.206.177|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12088943206 (11G) [application/x-gzip]
Saving to: ‘mpii_human_pose_v1.tar.gz.1’

mpii_human_pose_v1. 100%[===================>]  11.26G  18.2MB/s    in 10m 39s 

2021-03-27 19:35:11 (18.0 MB/s) - ‘mpii_human_pose_v1.tar.gz.1’ saved [12088943206/12088943206]



In [ ]:
!tar xzvf mpii_human_pose_v1.tar.gz

Streaming output truncated to the last 5000 lines.
./images/019279044.jpg
./images/068025069.jpg
./images/043439489.jpg
./images/065199085.jpg
./images/043180510.jpg
./images/077936327.jpg
./images/032166247.jpg
./images/085835606.jpg
./images/028663685.jpg
./images/085061606.jpg
./images/045571938.jpg
./images/052659494.jpg
./images/040132544.jpg
./images/041967229.jpg
./images/070550266.jpg
./images/006644163.jpg
./images/004278327.jpg
./images/061707653.jpg
./images/046332890.jpg
./images/038161650.jpg
./images/014551548.jpg
./images/012610270.jpg
./images/030097552.jpg
./images/090687283.jpg
./images/092028594.jpg
./images/020276921.jpg
./images/086201772.jpg
./images/081240248.jpg
./images/045623651.jpg
./images/002182340.jpg
./images/009557537.jpg
./images/039847876.jpg
./images/001289747.jpg
./images/062442493.jpg
./images/020601740.jpg
./images/023538326.jpg
./images/087342324.jpg
./images/061328639.jpg
./images/045247005.jpg
./images/043498418.jpg
./images/087985677.jpg
./imag

In [ ]:
import pandas as pd, csv
import os
import shutil

The MPII dataset annotations are stored in in a Matlab file structure, which makes it difficult to work with in Python and Jupyter Notebooks. Thank you to Github user Arifkhan21 (https://github.com/Arifkhan21/MPII-Dataset-in-CSV) for converting the annotations to a csv file.

In [118]:
dataset = pd.read_csv("https://raw.githubusercontent.com/Arifkhan21/MPII-Dataset-in-CSV/master/mpii_dataset.csv")

In [119]:
dataset

Unnamed: 0           NAME  ...     Activity        Category
0               0  015601864.jpg  ...      curling          sports
1               0  015599452.jpg  ...      curling          sports
2               0  005808361.jpg  ...      curling          sports
3               0  086617615.jpg  ...      curling          sports
4               0  060111501.jpg  ...      curling          sports
...           ...            ...  ...          ...             ...
17367           0  033474347.jpg  ...  pushing car  transportation
17368           0  082650067.jpg  ...  pushing car  transportation
17369           0  072772110.jpg  ...  pushing car  transportation
17370           0  039361034.jpg  ...  pushing car  transportation
17371           0  084761779.jpg  ...  pushing car  transportation

[17372 rows x 37 columns]

This dataset contains a lot of categories that don't involve much movement, and have obscured limbs. I'm removing like "inactivity quiet/light" as I don't expect them to be as useful. *Note: variables and files are named something along the lines of sports as the original idea was to just use the sports category.

In [126]:
Sports = set()
for i in dataset.index:
  if dataset["Category"][i] == "fishing and hunting" or dataset["Category"][i] == "home activities" or dataset["Category"][i] == "home repair" or dataset["Category"][i] == "inactivity quiet/light" or dataset["Category"][i] == "lawn and garden" or dataset["Category"][i] == "miscellaneous" or dataset["Category"][i] == "music playing" or dataset["Category"][i] == "occupation" or dataset["Category"][i] == "religious activities" or dataset["Category"][i] == "self care" or dataset["Category"][i] == "transportation" or dataset["Category"][i] == "volunteer activities":
    dataset.drop(i, inplace=True)
  else:
    Sports.add(dataset["NAME"][i])

In [127]:
dataset

Unnamed: 0  ... Category
0               0  ...   sports
1               0  ...   sports
2               0  ...   sports
3               0  ...   sports
4               0  ...   sports
...           ...  ...      ...
17311           0  ...   sports
17312           0  ...   sports
17313           0  ...   sports
17314           0  ...   sports
17315           0  ...   sports

[9002 rows x 37 columns]

In [128]:
len(dataset)

9002

9002 should be enough for now.
Save this to csv for later processing.

In [129]:
dataset.to_csv("sports_annotations_no_box.csv")

In [131]:
!rm -rf images_data

In [132]:
!mkdir images_data

In [133]:
for image in os.listdir('images'):
  if image in Sports:
    shutil.copy(f'images/{image}', 'images_data')

In [134]:
len(os.listdir('images_data'))

9002

Lastly, I download and save this data to Google Drive for easier access in other files/programs. These files will also be found in the respository.

In [135]:
from google.colab import drive

In [136]:
drive.mount('/content/drive')
with open(f'/content/drive/My Drive/CPEN 291 Datasets/Sports_Dataset/sports_annotations_no_box.csv', 'wb'):
  shutil.copy("sports_annotations_no_box.csv", "/content/drive/My Drive/CPEN 291 Datasets/Sports_Dataset/sports_annotations_no_box.csv")
drive.flush_and_unmount()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [138]:
!tar cfz images.tar.gz images_data/
drive.mount('/content/drive')

with open(f'/content/drive/My Drive/CPEN 291 Datasets/Sports_Dataset/images.tar.gz', 'wb'):
  shutil.copy('images.tar.gz', '/content/drive/My Drive/CPEN 291 Datasets/Sports_Dataset/images.tar.gz')
# for fn in Sports:
#   with open(f'/content/drive/My Drive/CPEN 291 Datasets/Sports_Dataset/Images/{fn}', 'wb'):
#     shutil.copy(f"images_data/{fn}", f"/content/drive/My Drive/CPEN 291 Datasets/Sports_Dataset/Images/{fn}")

drive.flush_and_unmount()

Mounted at /content/drive
